Splitting out stable subsets from `template-haskell` (`template-haskell-lift` and `template-haskell-quasiquote`) (#696)
ghc-proposals/ghc-proposals: proposals/0000-splitting-out-stable-interfaces-from-th.rst

As we expect GHC's internal AST to regularly evolve with each major version of GHC, it is likely that each new major release of GHC will force a new major release of the ``template-haskell`` library.

.. Note::
   In ``template-haskell-2.18``, a new field was added to the ``ConP`` constructor of ``Pat`` to express the possibility of a list of type applications as part of a constructor pattern.
   End-users then had to update their code to account for this change. ``yesod`` uses ``ConP`` in some code for generating typeclass instances.
   The code had to be changed to pass an extra ``[]`` argument. See: `the PR to yesod <https://github.com/yesodweb/yesod/pull/1754/files#diff-b0e5dbc5d4ca2998772f987cc5f27c5fc761b34549bdecc93892bbe142d89d26R30>`_.

When upgrading GHC, users are often also forced to upgrade to the new GHC bundled ``template-haskell`` library.

.. _why strategy:
Justifying our strategy
^^^^^^^^^^^^^^^^^^^^^^^
Our strategy is informed by the classes of usages of ``template-haskell`` found in the ecosystem. We can divide users as follows:

> * (A) Quote-and-splice clients: These users use only splices, quotes, ``DeriveLift`` or quasiquotes. These users might not even need to import the ``template-haskell`` library.
> * (B) Syntax-construction clients. These users construct Template Haskell syntax trees either directly through its constructors, or indirectly through the smart-constructors exported by ``Language.Haskell.TH.Lib``.
> * (C) Reification clients. These users, notably various forms of deriving, use reification to interrogate the program. Reification currently returns Template Haskell ASTs.
> * (D) Syntax-analysis clients. Some clients pattern match on Template Haskell syntax tree datatypes.

Reviewer: ```suggestion
* (\A) Quote-and-splice clients: These users use only splices, quotes, ``DeriveLift`` or quasiquotes. These users might not even need to import the ``template-haskell`` library.
* (\B) Syntax-construction clients. These users construct Template Haskell syntax trees either directly through its constructors, or indirectly through the smart-constructors exported by ``Language.Haskell.TH.Lib``.
* (\C) Reification clients. These users, notably various forms of deriving, use reification to interrogate the program. Reification currently returns Template Haskell ASTs.
* (\D) Syntax-analysis clients. Some clients pattern match on Template Haskell syntax tree datatypes.
```
This should fix the rendering of (A), (B), etc.

These diverse usages of the library lead to diverse levels of breakage when a new major version of ``template-haskell`` comes out. We can rank them from (A) with the least breakage to (D) with the most.
For instance, the ``uuid`` library, which just depends on ``template-haskell`` in order to provide a derived ``Lift`` instance (a type (A) client), in all likelyhood would only need to bump its upper-bound on the library.

> On the other-hand ``th-desugar``, which pattern matches on the entire syntax tree (a type (D) client), would have to make code changes on most releases of the library.

Reviewer: ```suggestion
On the other hand ``th-desugar``, which pattern matches on the entire syntax tree (a type (D) client), would have to make code changes on most releases of the library.
```

Type (A) users are already using interfaces which are quite stable. Yet, they have to update their upper bounds whenever they want to be compatible with a new major version of GHC.
The first concrete step in our strategy is to publish package that provide these stable APIs. We will return to the benefits of this in the next section.

(B-D) do not currently use stable subsets of the ``template-haskell`` interface.
In the future, we aim to continue this strategy, by identifying stable interfaces for these classes of users which aren't tightly coupled to the Template Haskell AST.
The smart-constructors from the ``Language.Haskell.TH.Lib`` module are a good starting point for type (B) clients. Another idea is to use smart-constructors based on the Haskell2010 AST (See: `GHC#20828 <https://gitlab.haskell.org/ghc/ghc/-/issues/20828>`_).
For type (C) clients, we can build on the existing `th-abstraction` library, and perhaps expose a refined AST that doesn't need to be as expressive as the surface language.

> Type (D) clients on the otherhand are likely to be difficult to accommodate, since they are inherently tightly coupled to the Template Haskell syntax trees.

Reviewer: ```suggestion
Type (D) clients on the other hand are likely to be difficult to accommodate, since they are inherently tightly coupled to the Template Haskell syntax trees.
```

Our strategy of splitting out stable subsets of the API has the advantage that it allows users to opt-in to more stability.
``template-haskell`` is used very widely in the ecosystem. This makes it important that any attempt to improve its stability doesn't force a change to all users.
Users who wish to continue to use ``template-haskell`` may continue to do so, and the interfaces will continue to be exposed in both the new packages and the old.

By focusing on smaller subsets of the API we also make it much easier to be compatible with multiple versions of GHC.
This is an important property for any stable package as it allows a user to upgrade their dependencies independently of GHC.
We plan to implement this by create compatibility shims using ``CPP`` or ``PatternSynonyms``.

.. _advantages:
Benefits of splitting out ``template-haskell-lift`` and ``template-haskell-quasiquote``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Publishing ``template-haskell-lift`` and ``template-haskell-quasiquote`` will be beneficial both for GHC and the ecosystem.

> The biggest benefit is that library authors who are just deriving or using ``Lift`` instances or just exposing ``Quasiquoter``s no longer need to depend on the entirety of ``template-haskell``.

Reviewer: ```suggestion
The biggest benefit is that library authors who are just deriving or using ``Lift`` instances or just exposing ``Quasiquoter``\s no longer need to depend on the entirety of ``template-haskell``.
```
This should fix the rendering of the code blocks.

These will be shipped with GHC.
They will also be buildable from Hackage.
They will be buildable with at a *minimum* the last 3 versions of GHC.

Their interfaces will be as follows:

``template-haskell-lift``::

   module TemplateHaskell.Lift
    ( Q
    , Code
    , Quote
    , Exp
    , Lift(..)
    , defaultLiftTyped -- a utility for writing `liftTyped` methods when an instance currently defines lift only
    , liftAddrCompat -- a utility for creating an `Addr#` value, eg, for defining `Lift ByteString`
    , liftIntCompat -- a utility for lifting an `Int` without causing issues when used with `OverloadedSyntax`
    )

``template-haskell-quasiquote``::

   module TemplateHaskell.Quasiquoter
    ( Q
    , Exp
    , Pat
    , Type
    , Dec
    , QuasiQuoter (QuasiQuoter, quoteExp, quotePat, quoteType, quoteDec)
    )

Note that these modules are in the ``TemplateHaskell.`` namespace rather than the ``Language.Haskell.TH.`` namespace.
The idea to use this less verbose namespace for the new stable interfaces is thanks to Adam Gundry.

Effect and Interactions
-----------------------

> This works towards removing the special case for ``template-haskell`` in (GR1), but on its own it doesn't achieve it.

Reviewer 2: What is (GR1)? Do you mean (A)?

Author: I've clarified this. It's referring to: https://github.com/ghc-proposals/ghc-proposals/blob/01c0d736919a25e2669296aa71dd58368b09ba72/principles.rst?plain=1#L359

But if these packages switch to depending on our new packages, then ``template-haskell`` could depend on them.
Currently ``template-haskell`` must vendor a small portion of ``filepath`` and ``containers``, and that would no longer be necessary.

Many boot packages depend on ``template-haskell``, but all of them only depend on it for the parts of the interface exposed by ``template-haskell-lift`` and ``template-haskell-quasiquote``.
If we can convince their maintainers to depend on these packages instead, then GHC would no longer (transitively) depend on ``template-haskell``.
This makes it possible for packages to depend on the ``ghc`` library at the same time as a version of ``template-haskell`` different to the one bundled with that GHC.

Proposed Change Specification
-----------------------------
No changes to the language or the compiler are required for this proposal.

Proposed Library Change Specification
-------------------------------------

We propose to publish two new libraries: ``template-haskell-lift`` and ``template-haskell-quasiquote``.
These will be shipped with GHC.
They will also be buildable from Hackage.

> They will be buildable with at a *minimum* the last 3 versions of GHC.

Author: TODO: flesh this out. In particular, we should mention that we will use CPP to shim over different versions of the interface.

The current version of the libraries are compatible with GHC 8.10 and later.

> 

Author: Mention what are dependencies are.
Mention that the library is reinstallable but shipped with GHC ala containers, bytestring, etc, but not baked in like ghc-internal.

----------------------
As this proposal deals exclusively with creating new packages, there are no backwards compatibility worries.

Alternatives
------------
The majority of the breaking changes to ``template-haskell`` comes from changes to the TH AST.
An alternative approach would be to simplify move the TH AST into a new package, and keep ``template-haskell``
as the remaining interface.

The main issue with this alternative is that it would force a change on basically all users but type (A).
This would be a large and wide ranging breaking change.

Another issue is that the ``Lift`` interface has changed much more frequently in the past than the ``Quasiquoter`` interface.
If either of these changed in the future, then every user would have to update their upper bounds. ]
Whereas with the split packages, you only need to update your bounds if the interface you actually depend on has changed.

Unresolved Questions
--------------------

- Should the modules live in the ```TemplateHaskell.`` or the ``Language.Haskell.TH.`` namespace?
- Should these packages live in the GHC repository, in another repository on the GHC Gitlab, or on GitHub?

Implementation Plan
-------------------

> Teo Camarasu has implemented a `MR <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13569>`_

Author: ```suggestion
Teo Camarasu has implemented an `MR <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13569>`_
```

On the other hand, the small interfaces exposed by ``template-haskell-lift`` and ``template-haskell-quasiquote`` are easy to make compatible with multiple versions of GHC.
They rarely change and if they don't change between two versions of GHC, then we can accommodate both for free.
If they do change, then it's likely that we can use ``CPP`` to expose to shim over GHC internals and expose a consistent interface.

.. note::
   For instance, `Overloaded Quotations proposal <./0246-overloaded-bracket.rst>`_ changed the type of the ``lift`` method of ``Lift`` from ``lift :: a -> Q a`` to ``lift :: Qoute m => a -> m a``.

   Suppose ``template-haskell-lift`` existed at the time and ``template-haskell-lift-0.1`` corresponded to the old interface and ``template-haskell-lift-0.2`` corresponded to the new interface.
   Further suppose that GHC-9.0 ships with ``template-haskell-lift-0.1`` and GHC-9.2 ships with and implements the interface of ``template-haskell-lift-0.2``.

   Our argument in this section is that it is convenient to make the following possible:

   * ``template-haskell-lift-0.1`` can be compiled with GHC-9.2

>    * ``template-haskell-lift-0.2`` can be compiled with GHC-9.0

Reviewer 3: I'm having difficulty following this example, unfortunately. Since `Lift` is a wired-in class, I don't see how `template-haskell-lift` can expose a version of it that uses a different `lift` method type than the one shipped by the compiler. Isn't it merely reexporting whatever `Lift` class the compiler provides?

The situation would be different if the class was exported without methods, of course. I don't know whether `DeriveLift` works if the class is in scope but the methods are not?

(Side note: the RST note rendering doesn't seem to work.)

Author: That's a good point. I should make this a bit clearer.

While we are forced to re-export `Lift` as is from `ghc-internal`, we do have a bit of leeway.

In one direction we export `type Quote m = Q ~ m` (see the [MR](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13569/diffs?file=79e359557bae9fa573412b8b1338fde5f60b608a#79e359557bae9fa573412b8b1338fde5f60b608a_0_34)). I haven't tried out the other direction, but I think we could define `class Quote` and make any instance of `Q` an instance of `Quote`.

So while the types wouldn't be exactly the same, it wouldn't be noticeable to end-users.

We couldn't do this for any arbitrary change, but I think it's likely that we could do it for most.

(I'll integrate this into the text)

Yeah I'm not sure what's going on with the RST note syntax. Perhaps it's just best if I drop it.

Author: On further thought, I think the direction I have on the MR is sound,
but I don't think the other direction is sound, since as you say we can't touch the class method. So an instance could use a method from `Q` not in `Quote`

Teo Camarasu is happy to take on any maintenance work necessary for these packages for the foreseeable future,
but someone else would have to take over if they are no longer able to.

Backward Compatibility
----------------------
As this proposal deals exclusively with creating new packages, there are no backwards compatibility worries.

Alternatives
------------
The majority of the breaking changes to ``template-haskell`` comes from changes to the TH AST.
An alternative approach would be to simplify move the TH AST into a new package, and keep ``template-haskell``
as the remaining interface.

The main issue with this alternative is that it would force a change on basically all users but type (A).
This would be a large and wide ranging breaking change.

Another issue is that the ``Lift`` interface has changed much more frequently in the past than the ``Quasiquoter`` interface.
If either of these changed in the future, then every user would have to update their upper bounds. ]
Whereas with the split packages, you only need to update your bounds if the interface you actually depend on has changed.

Unresolved Questions
--------------------

> - Should the modules live in the ```TemplateHaskell.`` or the ``Language.Haskell.TH.`` namespace?
> - Should these packages live in the GHC repository, in another repository on the GHC Gitlab, or on GitHub?

Author: ```suggestion
```
I'll drop these before merging

Concretely if ``template-haskell-0.1`` is shipped with GHC-9.14, then there should be minor releases (``template-haskell-0.1.*``) that can be built with GHC-9.12 and GHC-9.16.
We wish to have as broad a support range as feasible, eg, the current version of the libraries are compatible with GHC-8.10 up to GHC-9.12 (the present release).
We acknowledge that this might not always be possible, and that these interfaces might need to change in the future in ways that cannot be shimmed over.

These packages only depend on ``ghc-internal`` and ``base``. Crucially they do not depend on ``template-haskell``.

Their initial interfaces will be as follows:

``template-haskell-lift``::

   module TemplateHaskell.Lift
    ( Q
    , Code
    , Quote
    , Exp
    , Lift(..)
    , defaultLiftTyped -- a utility for writing `liftTyped` methods when an instance currently defines lift only
    , liftAddrCompat -- a utility for creating an `Addr#` value, eg, for defining `Lift ByteString`
    , liftIntCompat -- a utility for lifting an `Int` without causing issues when used with `OverloadedSyntax`
    )

``template-haskell-quasiquote``::

>    module TemplateHaskell.Quasiquoter

Reviewer 2: Yes, please make this change (`Q`).

These packages only depend on ``ghc-internal`` and ``base``. Crucially they do not depend on ``template-haskell``.

Their initial interfaces will be as follows:

``template-haskell-lift``::

   module TemplateHaskell.Lift
    ( Q
    , Code
    , Quote
    , Exp
    , Lift(..)
    , defaultLiftTyped -- a utility for writing `liftTyped` methods when an instance currently defines lift only
    , liftAddrCompat -- a utility for creating an `Addr#` value, eg, for defining `Lift ByteString`
    , liftIntCompat -- a utility for lifting an `Int` without causing issues when used with `OverloadedSyntax`
    )

``template-haskell-quasiquote``::

   module TemplateHaskell.Quasiquoter
    ( Q
    , Exp
    , Pat
    , Type
    , Dec
    , QuasiQuoter (QuasiQuoter, quoteExp, quotePat, quoteType, quoteDec)
    )

Note that these modules are in the ``TemplateHaskell.`` namespace rather than the ``Language.Haskell.TH.`` namespace.
The idea to use this less verbose namespace for the new stable interfaces is thanks to Adam Gundry.

Effect and Interactions
-----------------------

> This works towards removing the special case for ``template-haskell`` in `(GR1) <https://github.com/ghc-proposals/ghc-proposals/blob/master/principles.rst#33stability-gr1>`_ from `Principles for GHC <../principles.rst>`_, but on its own it doesn't achieve it.

Reviewer 2: (Non-actionable in this proposal.)

I just had a look at (GR1) and don't understand why `template-haskell` warrants a special case at all. Perhaps it used to when it was wired-in, but that's not the case any more since it nowadays re-exports from `ghc-internal`. The versioning of the latter accounts for the alluded breaking changes to GHC's AST.

Author: Yes we should reexamine GR1 and see if we can drop it

Whereas with the split packages, you only need to update your bounds if the interface you actually depend on has changed.

Just one stable package
^^^^^^^^^^^^^^^^^^^^^^^
This proposal splits out two interfaces from ``template-haskell`` into two packages.
An alternative would be to split them into one new package, perhaps called ``template-haskell-stable``.

This would be simpler as it would lead to fewer packages.
Yet, it would reduce the benefits from this proposal, especially if we choose to continue with this strategy of splitting out interfaces from ``template-haskell`` in the future.

`PVP <https://pvp.haskell.org/>`_ dictates that if any interface in a package changes in a breaking way, then the entire package needs to bump its major version.
The most unstable interface in a package determines the package's overall stability.
The more interfaces a package contains, the greater the risk of a frequent breaking changes.
If we choose to continue splitting interfaces out of the ``template-haskell`` package, this risk will grow further if we choose to put them into ``template-haskell-stable``.

> Eventually ``template-haskell-stable`` might no longer be stable at all.

Reviewer 4: The “more splitting means less spurious version bump” is always true. So we could always cut down the packages down to single function packages. I don't think the argument stands on its own. Maybe there's an optimal package size, but I don't know that we have a good way of characterising it in 2025.

The other arguments, I think, are more compelling. Maybe they should be evaluated in lights of the plans to have a better stage distinction (proposal [682](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0682-explicit-level-imports.rst)). @mpickering @alt-romes @adamgundry does the choice of splitting `Lift` into its own package interact at all with your plans? Or are they indifferent to this particular choice?

> Eventually ``template-haskell-stable`` might no longer be stable at all.

Reviewer 3: The argument for separate packages @TeofilC makes in https://github.com/ghc-proposals/ghc-proposals/pull/696#discussion_r2177107282 seems compelling to me. Fundamentally I think `Lift` and quasiquotes are very different features and bundling them together would be more confusing than helpful. Despite the name, quasiquotes aren't (TH) quotes at all, they are splices!

 > The other arguments, I think, are more compelling. Maybe they should be evaluated in lights of the plans to have a better stage distinction (proposal [682](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0682-explicit-level-imports.rst)). @mpickering @alt-romes @adamgundry does the choice of splitting `Lift` into its own package interact at all with your plans? Or are they indifferent to this particular choice?

From the point of view of keeping a strong stage distinction, `Lift` is best avoided where possible, because defining a `Lift` instance inherently forces the datatype involved to be available at both compile-time and runtime (see [section 5.3 of the Explicit Level Imports proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0682-explicit-level-imports.rst#implicit-lifting-and-deriving-lift-instances)). So I think it is potentially valuable for packages to be able to express "I define a quasiquoter, but not any `Lift` instances" via their dependencies.

> Eventually ``template-haskell-stable`` might no longer be stable at all.

Reviewer 5: > From the point of view of keeping a strong stage distinction, ...

Indeed in library providing the type, the quasiquoter definition would only need the data-type at"compile-time".

However a user of such quasiquoter would actually want to work with the type at "run-time".

So I think that any types "crossing"-through stage persistence will need to exist (i.e. the modules providing them) both at compile and run-time at some point of the pipeline.

TL;DR, I don't think that avoiding `Lift` instances actually saves much. I'd actually think that definining QuasiQuoters using `lift/liftTyped` is a good way to avoid silly mistakes (by writing ad-hoc lift). I.e. I think that library writers using `template-haskell-quasiquoters` should probably depend on `template-haskell-lift` too, but not necessarily vice-versa.

> Eventually ``template-haskell-stable`` might no longer be stable at all.

Reviewer 2: I think this section is quite long already. Furthermore it is unclear to me what further sentence could be added to improve the argument. 

To me the main point is "From the perspective of the client, the package version number induces a form of tight coupling on the package's API; change in one part of the API affects the shared version number of otherwise loosely coupled different parts of the API. This suggest that every coherent API subset should live in its own package. However, every additional package comes with maintenance overhead. Hence we initiate package splits based on evidence. There are many clients (around xx%) that only use `Lift` and clients that only define quasiquoters; hence these are two different packages to avoid artificial tight coupling.", but apparently YMMV.

This risk would be diminished if the majority of users used both the ``Lift`` and ``QuasiQuoter`` interfaces. In that case, users would need to update their bounds if either package had a major release,
so it would be equivalent to updating the bounds for ``template-haskell-stable``. But, this is not the case. The vast majority of the packages that depend on this interface only use ``Lift``.
Some use only ``QuasiQuoter``, and others use both.

We also be sceptical of a ``template-haskell-stable`` package because stability is not an essential property of an interface.
We can look back on the *past* stability of these interfaces, but we cannot know their *future* stability.
Part of the motivation of this proposal is to make it easier to accommodate future changes to these interfaces.
Our aim here isn't to split these interfaces out in order to fix them in stone, but to make it easier for end-users to cope with future changes,
and to eliminate unnecessary work when the subset of the interface they depend on hasn't actually changed.

Rather we should be looking at how interfaces are related. We should look at relations of tight coupling.

> If two interfaces are tightly coupled, it makes sense to group them into one package as changes to one will force changes to the other.

Reviewer 2: Ah, actually I wanted to delete this comment in favour of [my response](https://github.com/ghc-proposals/ghc-proposals/pull/696#discussion_r2189155137) on the other thread. I'll delete it.

So, their library might be built by an end-user with a version of GHC the maintainer hasn't yet tested.

We could argue that the maintainer doesn't have a reason to be worried since GHC comes with strong `stability principles <../principles.rst>`_. Yet, these also have exceptions.

Consider the following scenario:
1. GHC-9.14 adds support to ``DeriveLift`` for ``GADT``\s.
2. The maintainer adds ``DeriveLift`` instances for their ``GADT``\s.
3. The GHC maintainers realize that there is a fundamental issue with this new feature and that it must be reverted in GHC-9.16.
4. Now the maintainers code is broken by the release of GHC-9.16

A way to defend against this would be to count changes to ``DeriveLift`` as part of the interface ``template-haskell-lift``.
Then whenever the class of datatypes supported by ``DeriveLift`` changed, we would make a new major version of ``template-haskell-lift``.

While we hold these to be valid concerns, we do not think expanding the interface of ``template-haskell-lift`` is the way to resolve them.
This suggestion goes against our aim of decoupling ``template-haskell-lift`` from GHC.

> It is possible to shim over changes to the ``Lift`` interface in ``ghc-internal`` as we've outlined in :ref:`independence`.

Author: This reference works when I actually run sphinx but not on the github preview. I'm not sure if there's a better way to do internal links?

Reviewer 3: I've tripped over this limitation of GitHub rendering as well, and it's annoying. The only way I know is to link to the automatically-generated anchor, e.g.
```rst
`independence <#131upgrading-libraries-independently-of-ghc>`__
```
and hope the referenced section title doesn't change afterwards.

But it is not possible to shim over features implemented in the compiler.
If the implementation of ``DeriveLift`` changed then we would be forced to have a sequence of versions of ``template-haskell-lift`` without either backwards- or forwards-compatibility.
This would greatly limit the benefits of this proposal.

We feel that it is reasonable that a library maintainer should be able to explicitly define which versions of GHC they support, but that this can be done orthogonally to this proposal.
In fact, tackling this independently might lead to a more general solution.

> For instance, we could have a ``ghc-version`` package, which exports nothing but carries the version of GHC. We would have to allow it to be exempted from unused package warnings, but this seems doable.

Reviewer 2: (Non-actionable as far as the proposal goes.) 

Why isn't `ghc-version` simply `ghc-internal`? That is, we could just ensure that `ghc-internal` is versioned according to the GHC version it comes with. Which incidentally is the case *already*. So `ghc-internal` is `ghc-version`.
